scripts/data/src/org/akvo/gae/remoteapi/DeleteRepeatQuestionGroupIteration.java

```diff
@@ -0,0 +1,315 @@
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+public class DeleteRepeatQuestionGroupIteration implements Process {
+
+    private static long NON_EXISTENT_ID = -1L;
+
+    private Long surveyInstanceId;
+
+    private Long formId;
+
+    private String questionGroupName;
+
+    private List<Long> iterationsToDelete;
+
+    private boolean deleteIterations;
+
+    private boolean updateIterations;
+
+    @Override
+    public void execute(DatastoreService ds, String[] args) throws Exception {
+        System.out.println("##############################################################");
+        System.out.printf(
+                        "#      Arguments: survey instance id, current question group name between quotes, %n" +
+                        "#      comma-separated list of iterations (as they appear in data  %n" +
+                        "#      cleaning report i.e. starting at number 1...) to delete %n" +
+                        "#      d to confirm deletion of iterations %n");
+        System.out.printf("##############################################################%n%n%n");
+
+        parseArgs(args, ds);
+
+        List<Long> repeatGroupQuestionIds = retrieveRepeatGroupQuestionIds(ds);
+        if (repeatGroupQuestionIds.isEmpty()) {
+            System.out.println("No questions found for group: *" + questionGroupName + "*");
+            System.exit(0);
+        }
+
+        List<Entity> responses = retrieveResponses(ds, this.surveyInstanceId, repeatGroupQuestionIds);
+
+        if (deleteIterations) {
+            deleteRepeatGroupIterations(ds, responses, this.iterationsToDelete);
+        } else {
+            System.out.println("No delete request made");
+        }
+
+        if (updateIterations) {
+            updateResponseIterations(ds, responses, this.iterationsToDelete);
+        } else {
+            System.out.println("No iterations to update");
+        }
+    }
+
+    private void parseArgs(String[] args, DatastoreService ds) {
+        Long surveyInstanceId = parseSurveyInstanceId(args);
+        if (surveyInstanceId.equals(-1L)) {
+            System.out.println("No survey instance id found. Exiting");
+            System.exit(1);
+        } else {
+            this.surveyInstanceId = surveyInstanceId;
+        }
+
+        this.formId = retrieveFormId(surveyInstanceId, ds);
```

Comment:
what happens is formId is null?

Reply:
if we have only one question group matching then the data is deleted correctly but if for example we have two forms with the same question group, then we get an exception and cannot perform any updates.